### PR TITLE
ENH: `stats.variation`: add `axis` tuple support and shift to axis nan policy factory

### PR DESCRIPTION
**Pull request by tirthasheshpatel**
#### Reference issue

Towards #14651

#### What does this implement/fix?

This PR adds support for `axis` tuples to `scipy.stats.variation` and shifts `nan_policy` and `keepdims` handling to the `_axis_nan_policy_factory` decorator.

### CODE REVIEW

File: scipy/stats/_variation.py [hunk 1]
```diff
@@ -193,11 +197,11 @@ def variation(a, axis=0, nan_policy='propagate', ddof=0, *, keepdims=False):
         else:
             del shp[axis]
         if len(shp) == 0:
-            result = np.nan
+            result = np.asarray(np.nan)
```
Comment (mdhaber) [inline scipy/stats/_variation.py:200]: While we're at it, consider `_get_nan` to preserve the dtype, e.g.
https://github.com/scipy/scipy/blob/2f6008b0a3aea3ce9ce7ca85a0860dab28b7a5fd/scipy/stats/_stats_py.py#L524-L526

**Review by mdhaber — APPROVED** [review]
This all looks great. I'm just going to run some test locally and wait for tests to pass, then merge. Anything funny you noticed, or did the decorator basically "just work"?

**Comment by tirthasheshpatel** [comment]
> Anything funny you noticed, or did the decorator basically "just work"?

The decorator "just worked"! Since the function doesn't have much special behavior (like errors for small inputs), it looks like everything works as expected with the decorator.

File: scipy/stats/_variation.py (hunk 1 above)
Comment (tirthasheshpatel) [reply scipy/stats/_variation.py:200]: Sounds good, thanks for noticing that!

**Comment by mdhaber** [comment]
Erg. Looks like a real failure. Almost!

**Comment by mdhaber** [comment]
When CI is green, this looks good. It adds `axis` tuple and masked array support, and it adds extensive tests to the existing behaviors. 
Since there is something to fix, I would go ahead and use `_get_nan`. Please make sure that dtype is preserved in the second to last case:

```python
import numpy as np
from scipy import stats
from numpy.ma import masked_array as ma

fun = stats.variation

def try_case(x, message, kwds):
    try:
        print(message)
        res = fun(x, **kwds)
        print(res)
        print(type(res), res.dtype)
    except AttributeError:
        print(type(res))
    except Exception as e:
        print(f"{type(e)}: {str(e)}")

    try:
        print("vs (master)")
        res = fun(x, **kwds,_no_deco=True)
        print(res)
        print(type(res), res.dtype)
    except AttributeError:
        print(type(res))
    except Exception as e:
        print(f"{type(e)}: {str(e)}")

    print("----")

np.random.seed(0)
random_array = np.random.rand(4, 6)
random_mask = np.random.rand(4, 6) > 0.8
random_nan_array = random_array.copy()
random_nan_array[random_mask] = np.nan
random_ma = np.ma.masked_array(random_array, mask=random_mask)
random_ma32 = np.ma.masked_array(random_array, mask=random_mask, dtype=np.float32)

cases = [
          ([], 'Zero Observations:', {}),
          ([1.], 'One Observation:', {}),
          ([1., 2.], 'Two Observations', {}),
          (random_array, '2D Array, no axis', {}),
          (random_array, '2D Array, axis=0', {'axis': 0}),
          (random_array, '2D Array, axis=1', {'axis': 1}),
          (np.matrix(random_array), 'Matrix, axis=1', {'axis': 1}, ),
          (random_array, '2D Array, axis=None', {'axis': None}),
          (random_array, '2D Array, axis=(0,1)', {'axis': (0, 1)}),
          ([np.nan, 1., 2., 3.], '1D Array with NaN, No nan_policy', {}),
          ([np.nan, 1., 2., 3.], '1D Array with NaN, raise', {'nan_policy': 'raise'}),
          ([np.nan, 1., 2., 3.], '1D Array with NaN, propagate', {'nan_policy': 'propagate'}),
          ([np.nan, 1., 2., 3.], '1D Array with NaN, omit', {'nan_policy': 'omit'}),
          ([np.nan, np.nan, np.nan], '1D Array with All NaNs, omit', {'nan_policy': 'omit'}),
          ([np.nan, np.nan, 3.], '1D Array with all but one NaN, omit', {'nan_policy': 'omit'}),
          (ma([0., 1., 2., 3.], mask=[True, False, False, False]), '1D Masked array, one masked', {}),
          (ma([1., 2., 3.], mask=[True, True, True]), '1D Masked array, all masked', {}),
          (ma([1., 2., 3.], mask=[True, True, False]), '1D Masked array, all but one masked', {}),
          (random_nan_array, '2D Array with NaNs, propagate, axis=1', {'axis': 1, 'nan_policy': 'propagate'}, ),
          (random_nan_array, '2D Array with NaNs, omit, axis=1', {'axis': 1, 'nan_policy': 'omit'}, ),
          (random_ma, '2D Masked array, axis=1', {'axis': 1}, ),
          (random_array.astype(np.float32), '2D Array, float32, axis=1', {'axis': 1}),
          (random_nan_array.astype(np.float32), '2D Array with NaNs, float32, propagate, axis=1', {'axis': 1, 'nan_policy': 'propagate'}, ),
          (random_ma32, '2D Masked Array, float32, axis=1', {'axis': 1}, ),
          ]

for case in cases:
    try_case(*case)

```
Zero Observations:
nan
<class 'numpy.float64'> float64
vs (master)
nan
<class 'numpy.float64'> float64
----
One Observation:
0.0
<class 'numpy.float64'> float64
vs (master)
0.0
<class 'numpy.float64'> float64
----
Two Observations
0.3333333333333333
<class 'numpy.float64'> float64
vs (master)
0.3333333333333333
<class 'numpy.float64'> float64
----
2D Array, no axis
[0.21111021 0.09483777 0.56373451 0.56947454 0.6451295  0.17012705]
<class 'numpy.ndarray'> float64
vs (master)
[0.21111021 0.09483777 0.56373451 0.56947454 0.6451295  0.17012705]
<class 'numpy.ndarray'> float64
----
2D Array, axis=0
[0.21111021 0.09483777 0.56373451 0.56947454 0.6451295  0.17012705]
<class 'numpy.ndarray'> float64
vs (master)
[0.21111021 0.09483777 0.56373451 0.56947454 0.6451295  0.17012705]
<class 'numpy.ndarray'> float64
----
2D Array, axis=1
[0.15699989 0.33910148 0.89642052 0.20279467]
<class 'numpy.ndarray'> float64
vs (master)
[0.15699989 0.33910148 0.89642052 0.20279467]
<class 'numpy.ndarray'> float64
----
Matrix, axis=1
[0.15699989 0.33910148 0.89642052 0.20279467]
<class 'numpy.ndarray'> float64
vs (master)
[0.15699989 0.33910148 0.89642052 0.20279467]
<class 'numpy.ndarray'> float64
----
2D Array, axis=None
0.44380344970244456
<class 'numpy.float64'> float64
vs (master)
0.44380344970244456
<class 'numpy.float64'> float64
----
2D Array, axis=(0,1)
0.44380344970244456
<class 'numpy.float64'> float64
vs (master)
<class 'TypeError'>: '<' not supported between instances of 'tuple' and 'int'
----
1D Array with NaN, No nan_policy
nan
<class 'numpy.float64'> float64
vs (master)
nan
<class 'numpy.float64'> float64
----
1D Array with NaN, raise
<class 'ValueError'>: The input contains nan values
vs (master)
<class 'ValueError'>: The input contains nan values
----
1D Array with NaN, propagate
nan
<class 'numpy.float64'> float64
vs (master)
nan
<class 'numpy.float64'> float64
----
1D Array with NaN, omit
0.408248290463863
<class 'numpy.float64'> float64
vs (master)
0.408248290463863
<class 'numpy.float64'> float64
----
1D Array with All NaNs, omit
nan
<class 'numpy.float64'> float64
vs (master)
nan
<class 'numpy.float64'> float64
----
1D Array with all but one NaN, omit
0.0
<class 'numpy.float64'> float64
vs (master)
0.0
<class 'numpy.float64'> float64
----
1D Masked array, one masked
0.408248290463863
<class 'numpy.float64'> float64
vs (master)
0.7453559924999299
<class 'numpy.float64'> float64
----
1D Masked array, all masked
nan
<class 'numpy.float64'> float64
vs (master)
0.408248290463863
<class 'numpy.float64'> float64
----
1D Masked array, all but one masked
0.0
<class 'numpy.float64'> float64
vs (master)
0.408248290463863
<class 'numpy.float64'> float64
----
2D Array with NaNs, propagate, axis=1
[       nan 0.33910148        nan 0.20279467]
<class 'numpy.ndarray'> float64
vs (master)
[       nan 0.33910148        nan 0.20279467]
<class 'numpy.ndarray'> float64
----
2D Array with NaNs, omit, axis=1
[0.16734263 0.33910148 0.76664051 0.20279467]
<class 'numpy.ndarray'> float64
vs (master)
[0.16734263 0.33910148 0.76664051 0.20279467]
<class 'numpy.ndarray'> float64
----
2D Masked array, axis=1
[0.16734263 0.33910148 0.76664051 0.20279467]
<class 'numpy.ndarray'> float64
vs (master)
[0.15699989 0.33910148 0.89642052 0.20279467]
<class 'numpy.ndarray'> float64
----
2D Array, float32, axis=1
[0.15699987 0.33910143 0.8964205  0.20279467]
<class 'numpy.ndarray'> float32
vs (master)
[0.15699987 0.33910143 0.8964205  0.20279467]
<class 'numpy.ndarray'> float32
----
2D Array with NaNs, float32, propagate, axis=1
[       nan 0.33910143        nan 0.20279467]
<class 'numpy.ndarray'> float64
vs (master)
[       nan 0.33910143        nan 0.20279467]
<class 'numpy.ndarray'> float32
----
2D Masked Array, float32, axis=1
[0.16734263 0.33910143 0.7666405  0.20279467]
<class 'numpy.ndarray'> float32
vs (master)
[0.15699987 0.33910143 0.8964205  0.20279467]
<class 'numpy.ndarray'> float32
----

```
```

**Comment by mdhaber** [comment]
I think `entropy` is another easy candidate. The only special thing is that it will require a callable `n_samples`, but that is already supported by the decorator. `circmean`, `circvar`, and `circstd` also look like the should be easy.

**Comment by tirthasheshpatel** [comment]
> Please make sure that dtype is preserved in the second to last case

It isn't. I think because the decorator decides to put a float64 value. Without the decorator, the function behaves correctly and preserves dtype.

**Comment by tirthasheshpatel** [comment]
Should I make the `_get_nan` change in decorator too?

**Comment by mdhaber** [comment]
Oops, yeah. If the decorator doesn't use `_get_nan`, let's put that in there. Not absolutely essential in this PR, but would be great to do. LMK which you prefer.

**Comment by mdhaber** [comment]
Oh, and can the `keepdims` logic be removed from `variation` itself? Doesn't the decorator take care of it, assuming the function will always reduce?

**Comment by mdhaber** [comment]
I've tried resetting to the first commit, then removing all the `keepdims` and `nan_policy` stuff. I'm getting just one failure. When I clear that up, I'll submit a PR to your branch.

File: scipy/_lib/_util.py [hunk 2]
```diff
@@ -773,7 +773,11 @@ def _rng_spawn(rng, n_children):
 def _get_nan(*data):
     # Get NaN of appropriate dtype for data
     data = [np.asarray(item) for item in data]
-    dtype = np.result_type(*data, np.half)  # must be a float16 at least
+    try:
+        dtype = np.result_type(*data, np.half)  # must be a float16 at least
+    except np.exceptions.DTypePromotionError:
```
Comment (mdhaber) [inline scipy/_lib/_util.py:778]: What causes the `DTypePromotionError`?

Comment (tirthasheshpatel) [reply scipy/_lib/_util.py:778]: Mixed data types in test_gh16955

**Review by mdhaber — APPROVED** [review]
If CI is happy, this LGTM. You happy with it, too, @tirthasheshpatel? Any concerns? Looks like adding `_get_nan` to the decorator was pretty easy?

**Comment by tirthasheshpatel** [comment]
`_get_nan` seems to be working pretty well. I also confirmed with your script that the dtype is preserved now. So, LGTM too.

File: scipy/_lib/_util.py [hunk 3]
```diff
@@ -25,12 +25,14 @@
 
 if np.lib.NumpyVersion(np.__version__) >= '1.25.0':
     from numpy.exceptions import (
-        AxisError, ComplexWarning, VisibleDeprecationWarning  # noqa: F401
+        AxisError, ComplexWarning, VisibleDeprecationWarning,
+        DTypePromotionError # noqa: F401
```
[the remark this answers is not in the document]
Comment (tirthasheshpatel) [reply scipy/_lib/_util.py]: done, pretty cosmetic change, so didn't run the CI again.

**Pull request merged**